### DISCUSSION TITLE
importError

### DISCUSSION THREAD

**Discussion by selcuktopal80** [fetchai/cosmpy#359] [Q&A] [answered]
ImportError: cannot import name 'LedgerClient' from partially initialized module 'cosmpy.aerial.client' (most likely due to a circular import)

due to that error, not able to use or develop any fetchai git.

**Comment by jrriehl** [comment]
Can you please provide more context so that we can look into this? Some things that would help:
- What command are you running that results in this error?
- Is this the full error trace? If not, can you please provide it?
- What version of Python are you using?
- What is OS / environment are you running this in?

**Reply by selcuktopal80** [reply]
hi

just running python Script.py as usual:)

Python 3.10

Windows

here is the full error trace:

(venv) PS C:\Users\selcu\OneDrive\Masaüstü\cosmospy> python cosmos.py
Traceback (most recent call last):
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\cosmos.py", line 2, in <module>
    from cosmpy.aerial.client import LedgerClient, NetworkConfig
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\venv\lib\site-packages\cosmpy\aerial\client\__init__.py", line 32, in <module>
    from cosmpy.aerial.client.bank import create_bank_send_msg
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\venv\lib\site-packages\cosmpy\aerial\client\bank.py", line 23, in <module>
    from cosmpy.protos.cosmos.bank.v1beta1.tx_pb2 import MsgSend
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\venv\lib\site-packages\cosmpy\protos\cosmos\bank\v1beta1\tx_pb2.py", line 16, in <module>
    from cosmos.base.v1beta1 import coin_pb2 as cosmos_dot_base_dot_v1beta1_dot_coin__pb2
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\cosmos.py", line 2, in <module>
    from cosmpy.aerial.client import LedgerClient, NetworkConfig
ImportError: cannot import name 'LedgerClient' from partially initialized module 'cosmpy.aerial.client' (most likely due to a circular import) (C:\Users\selcu\OneDrive\Masaüstü\cosmospy\venv\lib\site-packages\cosmpy\aerial\client\__init__.py)

**Reply by jrriehl** [reply]
I think this happens because `cosmpy.aerial.client.bank` imports `Coin` from `cosmos.base.v1beta1.coin_pb2`, which you are importing in the same script as you import the `LedgerClient` from `cosmpy.aerial.client`. Are you only using `Coin` from `cosmos.base.v1beta1.coin_pb2`? If so, you could import this with:
```
from cosmos.aerial.client.bank import Coin
```
and that will hopefully eliminate the circular import error.

**Reply by jrriehl** [reply]
Actually, I might have misread this error. It looks like you are only importing `LedgerClient` and `NetworkConfig`.

I will try to reproduce this locally.

**Reply by jrriehl** [reply]
Are you importing anything else from `cosmpy` in the script `cosmos.py` either directly or in files that this script imports?

**Reply by selcuktopal80** [reply]
thx for the answer

my code:


from cosmpy.aerial.client import LedgerClient, NetworkConfig



ledger = LedgerClient(NetworkConfig.fetch_mainnet())

cfg = NetworkConfig(
    chain_id="cosmoshub-4",
    url="grpc+https://public-seed-node.cosmoshub.certus.one:26656",
    fee_minimum_gas_price=1,
    fee_denomination="uatom",
    staking_denomination="uatom",
)

ledger = LedgerClient(cfg)

balance = ledger.query_bank_balance('fetch12t9ewxmcawvh2q5hysgeezpp9gxfzzryta2k3t')

**Reply by selcuktopal80** [reply]
> I think this happens because `cosmpy.aerial.client.bank` imports `Coin` from `cosmos.base.v1beta1.coin_pb2`, which you are importing in the same script as you import the `LedgerClient` from `cosmpy.aerial.client`. Are you only using `Coin` from `cosmos.base.v1beta1.coin_pb2`? If so, you could import this with:
> 
> ```
> from cosmos.aerial.client.bank import Coin
> ```
> 
> and that will hopefully eliminate the circular import error.

i think it should be from cosmpy.aerial.client.bank import Coin, but not working

**Reply by selcuktopal80** [reply]
> Are you importing anything else from `cosmpy` in the script `cosmos.py` either directly or in files that this script imports?

even if i only do solely from cosmpy.aerial.client.bank import Coin , i recieve circular error again 

code:

from cosmpy.aerial.client.bank import Coin

ledger = LedgerClient(NetworkConfig.fetch_mainnet())

cfg = NetworkConfig(
    chain_id="cosmoshub-4",
    url="grpc+https://public-seed-node.cosmoshub.certus.one:26656",
    fee_minimum_gas_price=1,
    fee_denomination="uatom",
    staking_denomination="uatom",
)

ledger = LedgerClient(cfg)

balance = ledger.query_bank_balance('fetch12t9ewxmcawvh2q5hysgeezpp9gxfzzryta2k3t')


error: 

(venv) PS C:\Users\selcu\OneDrive\Masaüstü\cosmospy> python cosmos.py
Traceback (most recent call last):
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\cosmos.py", line 3, in <module>
    from cosmpy.aerial.client.bank import Coin
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\venv\lib\site-packages\cosmpy\aerial\client\__init__.py", line 32, in <module>
    from cosmpy.aerial.client.bank import create_bank_send_msg
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\venv\lib\site-packages\cosmpy\aerial\client\bank.py", line 23, in <module>
    from cosmpy.protos.cosmos.bank.v1beta1.tx_pb2 import MsgSend
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\venv\lib\site-packages\cosmpy\protos\cosmos\bank\v1beta1\tx_pb2.py", line 16, in <module>
    from cosmos.base.v1beta1 import coin_pb2 as cosmos_dot_base_dot_v1beta1_dot_coin__pb2
  File "C:\Users\selcu\OneDrive\Masaüstü\cosmospy\cosmos.py", line 3, in <module>
    from cosmpy.aerial.client.bank import Coin
ImportError: cannot import name 'Coin' from partially initialized module 'cosmpy.aerial.client.bank' (most likely due to a circular import) (C:\Users\selcu\OneDrive\Masaüstü\cosmospy\venv\lib\site-packages\cosmpy\aerial\client\bank.py)

**Comment by jrriehl** [comment, marked as answer]
So far I haven't been able to reproduce this error.

Can you confirm when you run exactly this code with CosmPy v0.8.0:

```python
from cosmpy.aerial.client import LedgerClient, NetworkConfig

ledger = LedgerClient(NetworkConfig.fetchai_mainnet())

balance = ledger.query_bank_balance('fetch12t9ewxmcawvh2q5hysgeezpp9gxfzzryta2k3t')
```

you get the same error?

**Reply by selcuktopal80** [reply]
now it works great, many thx @jrriehl

**Reply by jrriehl** [reply]
Great to hear! Let us know if you have any more issues.
